File: plugins/capawesome-capacitor-zip/android/src/main/java/io/capawesome/capacitorjs/plugins/zip/ZipPlugin.java
package io.capawesome.capacitorjs.plugins.zip;

import android.util.Log;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "Zip")
public class ZipPlugin extends Plugin {

    public static final String TAG = "Zip";
    public static final String ERROR_DESTINATION_MISSING = "destination must be provided.";
    public static final String ERROR_SOURCE_MISSING = "source must be provided.";
    public static final String ERROR_PASSWORD_REQUIRED = "password is required for encrypted files.";
    public static final String ERROR_PASSWORD_INCORRECT = "password is incorrect.";

    private Zip implementation;

    @Override
    public void load() {
        implementation = new Zip(this);
    }

    @PluginMethod
    public void unzip(PluginCall call) {
        try {
            String source = call.getString("source");
            if (source == null) {
                call.reject(ERROR_SOURCE_MISSING);
                return;
            }
            String destination = call.getString("destination");
            if (destination == null) {
                call.reject(ERROR_DESTINATION_MISSING);
                return;
            }
            String password = call.getString("password");

            implementation.unzip(source, destination, password);

            call.resolve();
        } catch (Exception exception) {
            String message = exception.getMessage();
            Log.e(TAG, message);
            call.reject(message);
        }
    }

    @PluginMethod
    public void zip(PluginCall call) {
        try {
            String source = call.getString("source");
            if (source == null) {
                call.reject(ERROR_SOURCE_MISSING);
                return;
            }
            String destination = call.getString("destination");
            if (destination == null) {
                call.reject(ERROR_DESTINATION_MISSING);
                return;
            }
            String password = call.getString("password");

            implementation.zip(source, destination, password);

            call.resolve();
        } catch (Exception exception) {
            String message = exception.getMessage();
            Log.e(TAG, message);
            call.reject(message);
        }
    }
}

File: plugins/capawesome-capacitor-zip/ios/Plugin/ZipPlugin.swift
import Foundation
import Capacitor

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(ZipPlugin)
public class ZipPlugin: CAPPlugin {
    public static let tag = "Zip"

    public let errorSourceMissing = "path must be provided."
    public let errorDestinationMissing = "path must be provided."
    public let errorOperationFailed = "Operation failed."

    private var implementation: Zip?

    override public func load() {
        self.implementation = Zip(plugin: self)
    }

    @objc func unzip(_ call: CAPPluginCall) {
        guard let source = call.getString("source") else {
            call.reject(errorSourceMissing)
            return
        }
        guard let destination = call.getString("destination") else {
            call.reject(errorDestinationMissing)
            return
        }
        let password = call.getString("password")

        implementation?.unzip(source: source, destination: destination, password: password, completion: { successful in
            if successful == true {
                call.resolve()
            } else {
                call.reject(self.errorOperationFailed)
            }
        })
    }

    @objc func zip(_ call: CAPPluginCall) {
        guard let source = call.getString("source") else {
            call.reject(errorSourceMissing)
            return
        }
        guard let destination = call.getString("destination") else {
            call.reject(errorDestinationMissing)
            return
        }
        let password = call.getString("password")

        implementation?.zip(source: source, destination: destination, password: password, completion: { successful in
            if successful == true {
                call.resolve()
            } else {
                call.reject(self.errorOperationFailed)
            }
        })
    }
}
